WordPress/wordcamp.org
# Add attendee survey to wordcamps.

Fixes #929 

This PR is an attempt at implementing the user flow described in this [comment](https://github.com/WordPress/wordcamp.org/issues/929#issuecomment-1654613966). 

##  How does this work programmatically?

- When the plugin is activated:
  - Create a survey page in draft status
  - Create a `tix_email` in draft status
  - Add Admin page to `central.wordpress.org`
  - Schedule a `daily` cron job that looks if camp ended 2 days ago.

- When the cron job runs:
  - Check if we already sent an email for this camp
  - Check to make sure the camp ended 2 days ago
  - Publish Survey page
  - Associate attendees to `tix_email`
  - Add `tix_email` to the queue (set to `pending`)
    - _email gets picked up by camptix cron email sending queue `tix_scheduled_every_ten_minutes`_
  - Schedule single event to unpublish the Survey page 2 weeks later.

## Screenshots

| View | Screenshot |
| --- | --- |
| Email queue | <img width="916" alt="Screenshot 2023-09-01 at 12 53 00 PM" src="https://github.com/WordPress/wordcamp.org/assets/1657336/84ebb44b-8778-4082-8376-0878bfb2243a"> |
| Email | <img width="626" alt="Screenshot 2023-08-28 at 3 11 52 PM" src="https://github.com/WordPress/wordcamp.org/assets/1657336/6d182c3c-156c-453d-8f14-478fcdecf4ce"> |
| Admin | <img width="1110" alt="Screenshot 2023-08-11 at 1 03 13 PM" src="https://github.com/WordPress/wordcamp.org/assets/1657336/e34e6412-4593-4b1e-9b4d-8b8ca77adca0"> |
| Survey Email | <img width="1063" alt="Screenshot 2023-08-11 at 10 55 18 AM" src="https://github.com/WordPress/wordcamp.org/assets/1657336/9374b20b-a64c-4aee-a6eb-7cce696d8adc"> |



## How to test

1. Create a Next-Gen Event WordCamp 
   - Make the Start Date in the past
   - Make the End Date the current date (we'll update this later) 
   - Create the site in the network    
     - Create a ticket: `wp-admin/edit.php?post_type=tix_ticket`
       - Make sure you update the Ticket Options. 
     - Setup Stripe for testing:
       - `wp-admin/edit.php?post_type=tix_ticket&page=camptix_options&tix_section=payment` 
         - You can you these keys:
           - Test Secret: `sk_test_4eC39HqLyjWDarjtT1zdp7dc` 
           - Test Publishable: `pk_test_TYooMQauvdEDq54NiTphI7jx`    
      - Publish the "tickets" page
      - Purchase a ticket using this test stripe credit card
        - Visa | 4242424242424242 | Any 3 digits | Any future date
1. Create a "traditional" WordCamp
   -  Create it in the network
1. Visit Next-Gen
   - Go to Pages
     - Expect to see `Attendee Survey` in `draft mode
     - Check out the content (if you don't have JetPack setup locally, you'll see and error when viewing the content, saying the JetPack form black is missing
   - Go to Ticket -> E-Mail
     - Expect to see an email in draft mode.
   - Run the following wp cli script to make sure the cron jobs are scheduled
     - `wp cron event list --url=https://events.wordpress.test/{city}/{year}/{type}` 
     - expect to see the following cron job: `wc_attendee_survey_email`
   - Go back to the Next-Gen Wordcamp Post type and update the End Date to 2 days ago.
   - Run this script to set WordCamp to closed `wp post update {post_id} --post_status=wcpt-closed --url=https://central.wordcamp.test/`
   - Run the following cli command:
     - run `wp cron event run wc_attendee_survey_email --url=https://events.wordpress.test/{city}/{year}/{type}` 
      - Should succeed 
    - run `wp cron event list --url=https://events.wordpress.test/{city}/{year}/{type}`
      - Should see a new scheduled task`wc_attendee_disable_survey`.
   - Visit  Ticket -> E-Mail -> `[[first_name], tell us what you thought: Post-event survey]`
     - It should be set to pending (or published if it has been picked up (the job runs every 10 mins )
     - If you are running the docker environment you can check out mailcatcher
       - `http://localhost:1080/`   
    - Visit the survey link on the front end an submit
    -  run `wp cron event run wc_attendee_disable_survey --url=https://events.wordpress.test/{city}/{year}/{type}` 
    - Revisit the link, the page should be unpublished.
    - Head over to `https://central.wordcamp.test/` click on "Attendee Survey"
      - Verify the list is accurate
 1. Make sure none of this exists on the traditional WordCamp :)
